lightning/src/ln/onion_utils.rs

```diff
@@ -965,34 +965,46 @@ where
 	}
 	let mut res: Option<FailureLearnings> = None;
 	let mut htlc_msat = *first_hop_htlc_msat;
-	let mut error_code_ret = None;
-	let mut error_packet_ret = None;
+	let mut _error_code_ret = None;
+	let mut _error_packet_ret = None;
 	let mut is_from_final_node = false;
 
 	const BADONION: u16 = 0x8000;
 	const PERM: u16 = 0x4000;
 	const NODE: u16 = 0x2000;
 	const UPDATE: u16 = 0x1000;
 
+	let mut onion_keys = Vec::with_capacity(path.hops.len());
+	construct_onion_keys_generic_callback(
+		secp_ctx,
+		&path.hops,
+		path.blinded_tail.as_ref(),
+		session_priv,
+		|shared_secret, _, _, route_hop_option: Option<&RouteHop>, _| {
+			onion_keys.push((route_hop_option.cloned(), shared_secret))
+		},
+	)
+	.expect("Route we used spontaneously grew invalid keys in the middle of it?");
+
 	// Handle packed channel/node updates for passing back for the route handler
-	let callback = |shared_secret, _, _, route_hop_opt: Option<&RouteHop>, route_hop_idx| {
+	for (route_hop_idx, (route_hop_option, shared_secret)) in onion_keys.into_iter().enumerate() {
```

Comment:
I like this refactoring. And I am also wondering if you can take it another step further by extracting this main for loop into a function that returns the `FailureLearnings`. To avoid this modification of variables outside of the loop.

Reply:
The conclusion here was that the function extract wasn't as easy as it seemed, right?